Add RoBERTa pretrain and finetune

tanhanzhuo opened the pull request:
<!-- Demo: https://github.com/PaddlePaddle/PaddleNLP/pull/26 -->
### PR types
<!-- One of [ New features | Bug fixes | Function optimization | Performance optimization | Breaking changes | Others ] -->
New features
### PR changes
<!-- One of [ Models | APIs | Docs | Others ] -->
 Docs
### Description
<!-- Describe what this PR does -->
本项目是RoBERTa模型在 Paddle 2.0上的开源实现，包含了数据tokenization和预训练代码。本项目旨在用简练清晰的代码完成基本预训练任务（仅Masked Language Model）。该代码易于理解，便于修改和定制。


## 对比Hugging Face

Hugging Face[预训练代码](https://github.com/huggingface/transformers/tree/master/examples/pytorch/language-modeling)和本目录下run_pretrain_roberta.py对比如下：

### 参数环境
二者使用的数据、超参、GPU保持一致：

训练数据：wikipedia-en

学习率、训练步长等参数见上面 预训练 章节

训练GPU：Tesla V100 32G * 2

### 运行速度

| 6 eopch | Hugging Face | PaddleNLP |
|---------|--------------|--------|
| FP16    |     90 h     |  72 h  | 
| FP32    |     154 h    |  139 h | 

FP32精度下，PaddleNLP较Hugging Face快11%左右。
FP16混合精度下，PaddleNLP较Hugging Face快25%左右

### GLUE任务结果
总训练tokens：512(seq_len）* 32(batch_size) * 780000(iteration)，约RoBERTa训练量10%，在GLUE validation set表现：

| Model GLUE Score   | CoLA  | SST-2  | MRPC   | STS-B  | QQP    | MNLI   | QNLI   | RTE    |
|--------------------|-------|--------|--------|--------|--------|--------|--------|--------|
| RoBERTa paper      |  68.0 |  96.4  |  90.9  |  92.4  |  92.2  |  90.2  |  94.7  |  86.6  |
|Hugging Face 6-epoch| 36.6  | 89.6   | 87.7   | 85.8   | 88.7   | 80.3   | 88.8   | 54.2   |
| PaddleNLP 6-epoch  | 36.9  | 89.5   | 84.3   | 86.2   | 88.6   | 80.5   | 88.4   | 58.1   |

整体来说，PaddleNLP和Hugging Face在Masked Language Modeling预训练中表现相似，因训练量较小，结果比原paper普遍低10%。另外，MRPC和RTE数据集validation set较小，结果波动幅度比较大，故PaddleNLP和Hugging Face二者结果有一定diff。

tanhanzhuo committed:
data collator for masked language modeling

tanhanzhuo committed:
tokenize data via HF datasets

tanhanzhuo committed:
start pertaining

tanhanzhuo committed:
Create README.md

tanhanzhuo committed:
Update README.md

ZeyuChen commented on examples/language_model/roberta/create_data.py:
```diff
@@ -0,0 +1,139 @@
+
+def main(args):
+    # Initialize the accelerator. We will let the accelerator handle device placement for us in this example.
+    accelerator = Accelerator()
```
不应该有这个

tanhanzhuo replied:
已去除

ZeyuChen commented on examples/language_model/roberta/README.md:
```diff
@@ -0,0 +1,57 @@
+# RoBERTa预训练（Masked Language Modeling）
+本项目是RoBERTa模型在 Paddle 2.0上的开源实现，包含了数据tokenization和预训练代码。本项目旨在用简练清晰的代码完成基本预训练任务（仅Masked Language Model）。该代码易于理解，便于修改和定制。
+## 快速开始
```
不需要快速开始
请按照数据准备，模型训练正常流程来呈现RoBERTa预训练。
然后再补上RoBERTa英语模型的finetune代码即可。

tanhanzhuo replied:
已修改

ZeyuChen reviewed (commented):
预训练仅作适宜，补充上finetune环节，使用RoBERTa英文模型的finetune几个简单任务作示意即可。

tanhanzhuo committed:
Update create_data.py

remove accelerator

tanhanzhuo committed:
Update create_data.py

tanhanzhuo committed:
Update run_glue.py

add RoBERTa Model

tanhanzhuo committed:
Update README.md

修改默认learning rate: 1e4 --> 3e5 （经测试，1e4在cola等数据集上容易造成训练不收敛）
增加tokenizer_name_or_path参数，便于用户使用不同path的model和tokenizer （原设定中只有model_name_or_path，要求model和tokenizer必须在同一个文件夹）

tanhanzhuo committed:
Update README.md

add fine-tune results
add comparison with Hugging Face

tanhanzhuo committed:
Update README.md

ZHUI commented on examples/benchmark/glue/run_glue.py:
```diff
@@ -248,7 +261,7 @@ def do_train(args):
 
     train_ds = load_dataset('glue', args.task_name, splits="train")
-    tokenizer = tokenizer_class.from_pretrained(args.model_name_or_path)
+    tokenizer = tokenizer_class.from_pretrained(args.tokenizer_name_or_path)
```
>  `tokenizer_name_or_path` 指示了Fine-tuning使用的具体tokenizer，一般保持和model_name_or_path一致，也可以单独指定

需要处理不设置的情况

tanhanzhuo replied:
已处理，没有设置tokenizer情况下使用model name

ZHUI commented on examples/language_model/roberta/README.md:
````diff
@@ -0,0 +1,126 @@
+
+```
+
+# 速度对比Hugging Face
````
感觉这部分可以放到 pr 的描述里面。这里主要目的是，验证训练是有效的即可，可以不用写到文档中。

ZHUI commented on examples/language_model/roberta/collator.py:
```diff
@@ -0,0 +1,68 @@
+# limitations under the License.
+
+
+import paddle
```
这块 DataCollator 与 @smallv0221 勾兑过了吗? 如商量过的话，可以放大 `paddlenlp/data/DataCollator` 这一类的位置。

tanhanzhuo replied:
目前datacollator主要是做padding等整合数据工作（'Stack', 'Pad', 'Tuple', 'Dict'），这里的collator则是整合数据并进行masking处理，可能和目前版本data collator目的不一致，所以也许不适合直接整合到collator里面？

ZHUI commented on examples/language_model/roberta/create_data.py:
```diff
@@ -0,0 +1,136 @@
+import os
+import random
+import datasets
+import torch
```
这里的 torch 依赖可否消除？

tanhanzhuo replied:
已消除

tanhanzhuo committed:
Update README.md

tanhanzhuo committed:
Update README.md

tanhanzhuo committed:
Update README.md

tanhanzhuo committed:
Update create_data.py

delete torch dependency

tanhanzhuo committed:
Update run_glue.py

if tokenizer_name not passed, use model_name

smallv0221 commented on examples/language_model/roberta/collator.py:
```diff
@@ -0,0 +1,68 @@
+import paddle
+from paddlenlp.data import Stack, Tuple, Pad, Dict
+class DataCollatorMLM():
+    def __init__(self, mask_token_id, pad_token_id, token_len, batch_pad=None):
```
这里参照HF的写法传个tokenizer进来吧

tanhanzhuo replied:
已修改

smallv0221 commented on examples/language_model/roberta/collator.py:
```diff
@@ -0,0 +1,68 @@
+    def __call__(self,examples):
+
+        examples = self.batch_pad(examples)
+        examples = [paddle.to_tensor(e) for e in examples]
```
这里可以不返回tensor么

smallv0221 reviewed (commented):
这个基本照着HF的DataCollatorForLanguageModeling写就可以，但是返回结果不要是tensor。

tanhanzhuo committed:
Rename run_pretrain_roberta.py to run_pretrain.py

tanhanzhuo committed:
Update README.md

tanhanzhuo committed:
Rename collator.py to utils.py

tanhanzhuo committed:
Update run_pretrain.py

collator -> utils

tanhanzhuo committed:
Update README.md

tanhanzhuo committed:
Update utils.py

mask_token_id, pad_token_id, token_len -> tokenizer as input

tanhanzhuo committed:
Update run_pretrain.py

mask_token_id, pad_token_id, token_len -> tokenizer

tanhanzhuo commented:
## 对比Hugging Face

Hugging Face[预训练代码](https://github.com/huggingface/transformers/tree/master/examples/pytorch/language-modeling)和本目录下run_pretrain_roberta.py对比如下：

### 参数环境
二者使用的数据、超参、GPU保持一致：

训练数据：wikipedia-en

学习率、训练步长等参数见上面 预训练 章节

训练GPU：Tesla V100 32G * 2

### 运行速度

| 6 eopch | Hugging Face | PaddleNLP |
|---------|--------------|--------|
| FP16    |     90 h     |  72 h  | 
| FP32    |     154 h    |  139 h | 

FP32精度下，PaddleNLP较Hugging Face快11%左右。
FP16混合精度下，PaddleNLP较Hugging Face快25%左右

### GLUE任务结果
总训练tokens：512(seq_len）* 32(batch_size) * 780000(iteration)，约RoBERTa训练量10%，在GLUE validation set表现：

| Model GLUE Score   | CoLA  | SST-2  | MRPC   | STS-B  | QQP    | MNLI   | QNLI   | RTE    |
|--------------------|-------|--------|--------|--------|--------|--------|--------|--------|
| RoBERTa paper      |  68.0 |  96.4  |  90.9  |  92.4  |  92.2  |  90.2  |  94.7  |  86.6  |
|Hugging Face 6-epoch| 36.6  | 89.6   | 87.7   | 85.8   | 88.7   | 80.3   | 88.8   | 54.2   |
| PaddleNLP 6-epoch  | 36.9  | 89.5   | 84.3   | 86.2   | 88.6   | 80.5   | 88.4   | 58.1   |

整体来说，PaddleNLP和Hugging Face在Masked Language Modeling预训练中表现相似，因训练量较小，结果比原paper普遍低10%。另外，MRPC和RTE数据集validation set较小，结果波动幅度比较大，故PaddleNLP和Hugging Face二者结果有一定diff。

tanhanzhuo committed:
Update README.md

delete HF comparison

ZeyuChen commented on examples/benchmark/glue/run_glue.py:
```diff
@@ -56,6 +57,7 @@
     "electra": (ElectraForSequenceClassification, ElectraTokenizer),
```
是否可以整体讲benchmark Auto化，这样无需每个模型都去改动代码。
@ZHUI 对你的Trainer脚本也应该是Auto化的体验，只让用户感知到模型名的变化，而无需开发带吗

ZHUI replied:
get

ZeyuChen commented on examples/benchmark/glue/run_glue.py:
```diff
@@ -88,6 +90,17 @@ def parse_args():
+        + ", ".join(
+            sum([
+                list(classes[-1].pretrained_init_configuration.keys())
+                for classes in MODEL_CLASSES.values()
```
如果整体Auto化的话，还需要这些复杂的配置吗？

ZeyuChen commented on examples/benchmark/glue/run_glue.py:
```diff
@@ -248,7 +261,10 @@ def do_train(args):
     train_ds = load_dataset('glue', args.task_name, splits="train")
-    tokenizer = tokenizer_class.from_pretrained(args.model_name_or_path)
+    if args.tokenizer_name_or_path:
+        tokenizer = tokenizer_class.from_pretrained(args.tokenizer_name_or_path) 
```
是否可以Auto化

tanhanzhuo replied:
已修改

ZeyuChen reviewed (commented):
glue脚本的finetune 尝试Auto化，而无需新增一个模型benchmark家一行MODEL CLASS和TOKENIZER CLASS

tanhanzhuo committed:
Update run_glue.py

switch to AutoClasses

tanhanzhuo committed:
Update README.md

delete model_type

tanhanzhuo commented:
修改成auto classes后，对比原先代码GLUE分数如下

Task/model | cola | sst2 | mrpc | stsb | rte
-- | -- | -- | -- | -- | --
Roberta-sep | 0.613 | 0.943 | 0.902 | 0.903 | 0.776
Roberta-auto | 0.591 | 0.944 | 0.905 | 0.903 | 0.776
Ernie-sep | 0.591 | 0.931 | 0.905 | 0.899 | 0.768
Ernie-auto | 0.594 | 0.931 | 0.903 | 0.899 | 0.761
Albert-sep | 0.537 | 0.922 | 0.741 | 0.177 | 0.552
Albert-auto | 0.538 | 0.913 | 0.676 | 0.129 | 0.451


所有实验使用的参数完全一致
其中*-sep代表原先需要指定model_type的实验结果
*-auto则为使用AutoClasses （AutoTokenizer, AutoModelForSequenceClassification）结果
二者结果基本相同（Albert可能需要特殊的调参技巧，默认参数结果很差，实验对比结果价值不高）

tanhanzhuo committed:
Merge branch 'develop' into roberta_pretrain

tanhanzhuo commented:
无法固定random seed问题，在设置了random.seed(args.seed),np.random.seed(args.seed),paddle.seed(args.seed)之后，仍然无法固定训练过程和结果。以下为三次实验结果（参数完全一致）：

run | cola | sst2 | mrpc | stsb | rte
-- | -- | -- | -- | -- | --
1 | 0.3721270886368030 | 0.8967889908256880 | 0.8452998846597460 | 0.8622340123463550 | 0.5812274368231040
2 | 0.3419338336781660 | 0.8956422018348620 | 0.8441728804197730 | 0.8622335101221840 | 0.5812274368231040
3 | 0.3809615036075180 | 0.8933486238532110 | 0.8367034313725490 | 0.8622336049503720 | 0.5812274368231040
Max diff | 0.0390276699293520 | 0.0034403669724770 | 0.0085964532871971 | 0.0000005022241710 | 0.0000000000000000


这可能会导致当前example目录下，示例中的实验结果无法复现。
Hugging face中，固定random seed后可以完全复现训练过程、下游任务结果。

tanhanzhuo committed:
style check

tanhanzhuo committed:
style check3

tanhanzhuo committed:
run_glue with model type

ZHUI commented on examples/language_model/roberta/create_data.py:
```diff
@@ -0,0 +1,179 @@
+import random
+import datasets
+from datasets import load_dataset
+from torch.utils.data import DataLoader
```
del

ZHUI commented on examples/language_model/roberta/run_pretrain.py:
```diff
@@ -0,0 +1,184 @@
+    # model = RobertaForMaskedLM(
+    #     RobertaModel(**RobertaForMaskedLM.pretrained_init_configuration[
+    #         args.model_name_or_path])) 
+    model = RobertaForMaskedLM(RobertaModel(**roberta_arch))
```
check

ZHUI committed:
Update create_data.py

ZHUI reviewed (approved):
LGTM

ZHUI committed:
Merge branch 'develop' into roberta_pretrain

ZHUI committed:
Merge branch 'develop' into roberta_pretrain

ZHUI committed:
Merge branch 'develop' into roberta_pretrain

ZHUI committed:
Merge branch 'develop' into roberta_pretrain
